lib/spack/spack/directory_layout.py

```diff
@@ -251,6 +251,50 @@ def build_env_path(self, spec):
     def build_packages_path(self, spec):
         return os.path.join(self.metadata_path(spec), self.packages_dir)
 
+    def _mkdirp(self, absolute_path, mode):
+        """
+        This mkdirp wrapper sets permissions to non-existing nested folders.
+
+        Nested folders are the ones which absolute_path prefix
+        is self.root (excluded)
+
+        e.g. absolute_path=self.root/a/b/c
+            If a path in set {a,b,c} does not exists, it will create it and
+            set permissions, otherwise it doesn't even touch it.
+            If self.root/a already exists, it will just create and
+            set permissions for b and c
+
+        Argument:
+            - absolute_path: (prefix must be be self.root)
+                absolute path of the folder hierarcy to create
+            - mode:
+                permissions to set to all non existing nested folders
```

Comment:
We need to think more carefully about this.

If we call `spack install foo` when the permissions for `foo` are relatively restrictive, and `bar` is already installed, we do not want to make the directory permissions for the architecture and compiler directories more restrictive.

I think what we need to do is set the intermediate directory permissions to be as permissive as possible. It's possible that we need to add a configuration option for those, but I hope not.

Reply:
Good point.

Unfortunately, I don't have a full overview of the security about packages, but I agree with you with the proposal of having an "open" hierarchy and setting permissions just to the package folder.

And as you are pointing out, how much "open" may be defined in configuration. But maybe it would be ok to just set right access restriction to the install folder (e.g. opt/spack) during spack setup and using the global read/write permissions defined in **packages.yaml/packages:all:permissions**.

Reply:
That wouldn't be sufficient. The global permissions ``packages:all:permissons`` could be more restrictive than the permissions for a specific package.

I think what we really want is for the intermediate directories to be world readable. There is no data stored in those directories, and the `install_tree` directory (default `$spack/opt/spack`) provides an easy chokepoint to restrict access if the names of the installed packages themselves are sensitive.

Other than that, I don't think we need to restrict the intermediate directory access at all since all install prefixes are controlled by their install permissions, and as I said no data is stored at the intermediate levels.

@albestro would that work for you?

Reply:
Yeah, you're right.

I agree with you with the fact that `$spack/opt/spack` is a chokepoint to control the security of the entire `install_dir`.

I'm also with you for what concerns `install_tree` directory openness, so everyone that has rights can install new packages in it, and at the same time the package security is based solely on the permissions of each package folder (this latter one controllable from configuration).

So if I understood correctly, it is ok for you setting 0o777 for intermediate folders under `$spack/opt/spack` and setting individual mode just for the leaf folder (package folder).


Moreover, with the aim to address what @tldahlgren and you were suggesting about function location, in the next commit I will remove the wrapper and extend `llnl/util/filesystem.py:mkdirs` function.